File: app/src/main/java/com/theandroiddev/mywins/data/db/DBAdapter.java
package com.theandroiddev.mywins.data.db;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.util.Log;

import com.theandroiddev.mywins.data.models.Success;
import com.theandroiddev.mywins.data.models.SuccessImage;
import com.theandroiddev.mywins.utils.Constants;

import java.util.ArrayList;

import static com.theandroiddev.mywins.utils.Constants.ADD_ON_TOP;
import static com.theandroiddev.mywins.utils.Constants.CATEGORY;
import static com.theandroiddev.mywins.utils.Constants.CATEGORY_VALUE;
import static com.theandroiddev.mywins.utils.Constants.DATE_ENDED;
import static com.theandroiddev.mywins.utils.Constants.DATE_ENDED_VALUE;
import static com.theandroiddev.mywins.utils.Constants.DATE_STARTED;
import static com.theandroiddev.mywins.utils.Constants.DATE_STARTED_VALUE;
import static com.theandroiddev.mywins.utils.Constants.DESCRIPTION;
import static com.theandroiddev.mywins.utils.Constants.DESCRIPTION_VALUE;
import static com.theandroiddev.mywins.utils.Constants.IMAGE_ID;
import static com.theandroiddev.mywins.utils.Constants.IMAGE_PATH;
import static com.theandroiddev.mywins.utils.Constants.IMPORTANCE;
import static com.theandroiddev.mywins.utils.Constants.IMPORTANCE_VALUE;
import static com.theandroiddev.mywins.utils.Constants.SUCCESS_ID;
import static com.theandroiddev.mywins.utils.Constants.SUCCESS_ID_VALUE;
import static com.theandroiddev.mywins.utils.Constants.TB_NAME_IMAGES;
import static com.theandroiddev.mywins.utils.Constants.TB_NAME_SUCCESSES;
import static com.theandroiddev.mywins.utils.Constants.TITLE;
import static com.theandroiddev.mywins.utils.Constants.TITLE_VALUE;
import static com.theandroiddev.mywins.utils.Constants.dummyCategory;
import static com.theandroiddev.mywins.utils.Constants.dummyDescription;
import static com.theandroiddev.mywins.utils.Constants.dummyEndDate;
import static com.theandroiddev.mywins.utils.Constants.dummyImportance;
import static com.theandroiddev.mywins.utils.Constants.dummyStartDate;
import static com.theandroiddev.mywins.utils.Constants.dummySuccessesSize;
import static com.theandroiddev.mywins.utils.Constants.dummyTitle;

/**
 * Created by jakub on 14.08.17.
 */

public class DBAdapter {
    private static final String TAG = "DBAdapter";
    private Context context;
    private SQLiteDatabase sqLiteDatabase;
    private DBHelper dbHelper;

    public DBAdapter(Context context) {

        this.context = context;
        dbHelper = new DBHelper(context);
        openDB();

    }

    public void openDB() {

        sqLiteDatabase = dbHelper.getWritableDatabase();

    }

    public void closeDB() {
        dbHelper.close();
    }

    public void addSuccess(Success success) {

        sqLiteDatabase.insert(TB_NAME_SUCCESSES, SUCCESS_ID, getSuccessContentValues(success));
    }

    public Cursor retrieveSuccess(String searchTerm, String sort) {

        String[] columns = {SUCCESS_ID, TITLE, CATEGORY, IMPORTANCE, DESCRIPTION, DATE_STARTED, DATE_ENDED};
        Cursor cursor;
        if (searchTerm != null && searchTerm.length() > 0) {
            String sql = "SELECT * FROM " + TB_NAME_SUCCESSES + " WHERE " + TITLE + " LIKE '%" + searchTerm + "%'";
            cursor = sqLiteDatabase.rawQuery(sql, null);
            return cursor;
        }
        cursor = sqLiteDatabase.query(TB_NAME_SUCCESSES, columns, null, null, null, null, sort);
        return cursor;
    }

    public void removeSuccess(ArrayList<Success> successesToRemove) {


        for (int i = 0; i < successesToRemove.size(); i++) {
            sqLiteDatabase.delete(TB_NAME_SUCCESSES, SUCCESS_ID + "=? and " + TITLE + "=?", new String[]{String.valueOf(successesToRemove.get(i).getId()), successesToRemove.get(i).getTitle()});

        }

    }

    public void editSuccess(Success showSuccess) {

        sqLiteDatabase.update(TB_NAME_SUCCESSES, getSuccessContentValues(showSuccess), SUCCESS_ID + "=" + showSuccess.getId(), null);
    }

    private void addSuccessImages(ArrayList<SuccessImage> successImage) {

        for (int i = 1; i < successImage.size(); i++) {

            ContentValues contentValues = new ContentValues();
            contentValues.put(IMAGE_PATH, successImage.get(i).getImagePath());
            contentValues.put(SUCCESS_ID, successImage.get(i).getSuccessId());
            sqLiteDatabase.insert(TB_NAME_IMAGES, IMAGE_ID, contentValues);

        }

    }

    public ArrayList<SuccessImage> getSuccessImages(int successId) {

        ArrayList<SuccessImage> successImages = new ArrayList<>();

        String sql = "SELECT * FROM " + TB_NAME_IMAGES + " WHERE " + SUCCESS_ID + "=" + successId;
        Cursor cursor = sqLiteDatabase.rawQuery(sql, null);
        if (cursor.moveToFirst()) {
            do {
                SuccessImage sI = new SuccessImage(successId);
                sI.setImagePath(cursor.getString(1));

                successImages.add(sI);

            } while (cursor.moveToNext());
        }
        cursor.close();
        return successImages;
    }

    public void editSuccessImages(ArrayList<SuccessImage> successImages, int successId) {

        deleteImages(successId);
        addSuccessImages(successImages);

    }

    private void deleteImages(int successId) {
        sqLiteDatabase.execSQL("DELETE FROM " + TB_NAME_IMAGES + " WHERE " + SUCCESS_ID + " = " + successId);

    }

    public ArrayList<Success> getDefaultSuccesses() {

        ArrayList<Success> successList = new ArrayList<>();
        new Constants();

        for (int i = 0; i < dummySuccessesSize; i++) {
            successList.add(new Success(dummyTitle.get(i), dummyCategory.get(i), dummyImportance.get(i), dummyDescription.get(i),
                    dummyStartDate.get(i), dummyEndDate.get(i)));
            Log.e(TAG, "insertDummyData: " + i);
        }

        return successList;
    }

    public Success getSuccess(int id) {

        String sql = "SELECT * FROM " + TB_NAME_SUCCESSES + " WHERE " + SUCCESS_ID + " = " + id;
        Cursor cursor = sqLiteDatabase.rawQuery(sql, null);

        if (cursor.moveToFirst()) {
            String title = cursor.getString(1);
            String category = cursor.getString(2);
            int importance = cursor.getInt(3);
            String description = cursor.getString(4);
            String dateStarted = cursor.getString(5);
            String dateEnded = cursor.getString(6);
            cursor.close();
            Success s = new Success(title, category, importance, description, dateStarted, dateEnded);
            s.setId(id);
            s.setId(id);
            return s;
        }


        return null;
    }

    public ArrayList<Success> getSuccesses(String searchTerm, String sortType,
                                           boolean isSortingAscending) {

        Success success;
        ArrayList<Success> successList = new ArrayList<>();
        Cursor cursor = retrieveSuccess(searchTerm, sortType);

        while (cursor.moveToNext()) {
            int id = cursor.getInt(SUCCESS_ID_VALUE);
            String title = cursor.getString(TITLE_VALUE);
            String category = cursor.getString(CATEGORY_VALUE);
            int importance = cursor.getInt(IMPORTANCE_VALUE);
            String description = cursor.getString(DESCRIPTION_VALUE);
            String dateStarted = cursor.getString(DATE_STARTED_VALUE);
            String dateEnded = cursor.getString(DATE_ENDED_VALUE);

            success = new Success(title, category, importance, description, dateStarted, dateEnded);
            success.setId(id);
            if (isSortingAscending) {
                successList.add(ADD_ON_TOP, success);
            } else {
                successList.add(success); //default = add on bottom
            }

        }

        cursor.close();

        return successList;
    }

    private ContentValues getSuccessContentValues(Success success) {
        ContentValues contentValues = new ContentValues();
        contentValues.put(TITLE, success.getTitle());
        contentValues.put(CATEGORY, success.getCategory());
        contentValues.put(IMPORTANCE, success.getImportance());
        contentValues.put(DESCRIPTION, success.getDescription());
        contentValues.put(DATE_STARTED, success.getDateStarted());
        contentValues.put(DATE_ENDED, success.getDateEnded());
        return contentValues;
    }


}

File: app/src/main/java/com/theandroiddev/mywins/data/db/DBAdapter.kt
package com.theandroiddev.mywins.data.db

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import com.theandroiddev.mywins.data.models.Success
import com.theandroiddev.mywins.data.models.SuccessImage
import com.theandroiddev.mywins.utils.Constants
import com.theandroiddev.mywins.utils.Constants.Companion.ADD_ON_TOP
import com.theandroiddev.mywins.utils.Constants.Companion.CATEGORY
import com.theandroiddev.mywins.utils.Constants.Companion.CATEGORY_VALUE
import com.theandroiddev.mywins.utils.Constants.Companion.DATE_ENDED
import com.theandroiddev.mywins.utils.Constants.Companion.DATE_ENDED_VALUE
import com.theandroiddev.mywins.utils.Constants.Companion.DATE_STARTED
import com.theandroiddev.mywins.utils.Constants.Companion.DATE_STARTED_VALUE
import com.theandroiddev.mywins.utils.Constants.Companion.DESCRIPTION
import com.theandroiddev.mywins.utils.Constants.Companion.DESCRIPTION_VALUE
import com.theandroiddev.mywins.utils.Constants.Companion.IMAGE_ID
import com.theandroiddev.mywins.utils.Constants.Companion.IMAGE_PATH
import com.theandroiddev.mywins.utils.Constants.Companion.IMPORTANCE
import com.theandroiddev.mywins.utils.Constants.Companion.IMPORTANCE_VALUE
import com.theandroiddev.mywins.utils.Constants.Companion.SUCCESS_ID
import com.theandroiddev.mywins.utils.Constants.Companion.SUCCESS_ID_VALUE
import com.theandroiddev.mywins.utils.Constants.Companion.TB_NAME_IMAGES
import com.theandroiddev.mywins.utils.Constants.Companion.TB_NAME_SUCCESSES
import com.theandroiddev.mywins.utils.Constants.Companion.TITLE
import com.theandroiddev.mywins.utils.Constants.Companion.TITLE_VALUE
import com.theandroiddev.mywins.utils.Constants.Companion.dummyCategory
import com.theandroiddev.mywins.utils.Constants.Companion.dummyDescription
import com.theandroiddev.mywins.utils.Constants.Companion.dummyEndDate
import com.theandroiddev.mywins.utils.Constants.Companion.dummyImportance
import com.theandroiddev.mywins.utils.Constants.Companion.dummyStartDate
import com.theandroiddev.mywins.utils.Constants.Companion.dummyTitle
import java.util.*
import javax.inject.Inject

/**
 * Created by jakub on 14.08.17.
 */

class DBAdapter @Inject constructor(
        val dbHelper: DBHelper
) {
    private var sqLiteDatabase: SQLiteDatabase? = null

    val defaultSuccesses: ArrayList<Success> = arrayListOf()

    init {

        openDB()
        Constants()
        var i = 0
        while (i < 5) {
            defaultSuccesses.add(Success(null, dummyTitle[i], dummyCategory[i], dummyDescription[i],
                    dummyStartDate[i], dummyEndDate[i], dummyImportance[i]))
            i++
        }
    }

    fun openDB() {

        sqLiteDatabase = dbHelper.writableDatabase

    }

    fun closeDB() {
        //        dbHelper.close();
    }

    fun addSuccess(success: Success) {

        sqLiteDatabase?.insert(TB_NAME_SUCCESSES, SUCCESS_ID, getSuccessContentValues(success))
    }

    private fun retrieveSuccess(searchTerm: String?, sortType: String?): Cursor? {

        if (sqLiteDatabase != null) {
            val columns = arrayOf(SUCCESS_ID, TITLE, CATEGORY, IMPORTANCE, DESCRIPTION, DATE_STARTED, DATE_ENDED)
            return if (searchTerm != null && searchTerm.isNotEmpty()) {
                val sql = "SELECT * FROM $TB_NAME_SUCCESSES WHERE $TITLE LIKE '%$searchTerm%'"
                sqLiteDatabase?.rawQuery(sql, null)
            } else {
                sqLiteDatabase?.query(TB_NAME_SUCCESSES, columns, null, null, null, null, sortType)
            }
        } else return null

    }

    fun removeSuccess(successesToRemove: ArrayList<Success>) {


        for (i in successesToRemove.indices) {
            sqLiteDatabase?.delete(TB_NAME_SUCCESSES, "$SUCCESS_ID=? and $TITLE=?", arrayOf(successesToRemove[i].id.toString(), successesToRemove[i].title))

        }

    }

    fun editSuccess(showSuccess: Success) {

        sqLiteDatabase?.update(TB_NAME_SUCCESSES, getSuccessContentValues(showSuccess), SUCCESS_ID + " = " + showSuccess.id, null)
    }

    private fun addSuccessImages(successImage: ArrayList<SuccessImage>) {

        for (image in successImage) {

            val contentValues = ContentValues()
            contentValues.put(IMAGE_PATH, image.imagePath)
            contentValues.put(SUCCESS_ID, image.successId)
            sqLiteDatabase?.insert(TB_NAME_IMAGES, IMAGE_ID, contentValues)

        }

    }

    fun getSuccessImages(successId: Long?): ArrayList<SuccessImage> {

        val successImages = ArrayList<SuccessImage>()

        val sql = "SELECT * FROM $TB_NAME_IMAGES WHERE $SUCCESS_ID=$successId"
        val cursor = sqLiteDatabase?.rawQuery(sql, null)
        if (cursor != null) {
            if (cursor.moveToFirst()) {
                do {
                    val sI = SuccessImage(null, successId, cursor.getString(1))

                    successImages.add(sI)

                } while (cursor.moveToNext())
            }
            cursor.close()
        }
        return successImages
    }

    fun editSuccessImages(successImages: ArrayList<SuccessImage>, successId: Long?) {

        deleteImages(successId)
        addSuccessImages(successImages)

    }

    private fun deleteImages(successId: Long?) {
        sqLiteDatabase?.execSQL("DELETE FROM $TB_NAME_IMAGES WHERE $SUCCESS_ID = $successId")

    }

    fun fetchSuccess(id: Long?): Success? {

        val sql = "SELECT * FROM $TB_NAME_SUCCESSES WHERE $SUCCESS_ID = $id"
        val cursor = sqLiteDatabase?.rawQuery(sql, null)

        if (cursor != null) {
            if (cursor.moveToFirst()) {
                val title = cursor.getString(1)
                val category = cursor.getString(2)
                val importance = cursor.getInt(3)
                val description = cursor.getString(4)
                val dateStarted = cursor.getString(5)
                val dateEnded = cursor.getString(6)
                cursor.close()
                val s = Success(id, title, category, description, dateStarted, dateEnded, importance)
                s.id = id
                //s.setId(id);
                return s
            }
        }

        return null
    }

    fun getSuccesses(searchTerm: String?, sortType: String?,
                     isSortingAscending: Boolean): ArrayList<Success> {

        var success: Success
        val successList = ArrayList<Success>()
        val cursor = retrieveSuccess(searchTerm, sortType)

        if (cursor != null) {
            while (cursor.moveToNext()) {
                val id = cursor.getLong(SUCCESS_ID_VALUE)
                val title = cursor.getString(TITLE_VALUE)
                val category = cursor.getString(CATEGORY_VALUE)
                val importance = cursor.getInt(IMPORTANCE_VALUE)
                val description = cursor.getString(DESCRIPTION_VALUE)
                val dateStarted = cursor.getString(DATE_STARTED_VALUE)
                val dateEnded = cursor.getString(DATE_ENDED_VALUE)

                success = Success(id, title, category, description, dateStarted, dateEnded, importance)
                if (isSortingAscending) {
                    successList.add(ADD_ON_TOP, success)
                } else {
                    successList.add(success) //default = add on bottom
                }

            }
            cursor.close()
        }

        return successList
    }

    private fun getSuccessContentValues(success: Success): ContentValues {
        val contentValues = ContentValues()
        contentValues.put(TITLE, success.title)
        contentValues.put(CATEGORY, success.category)
        contentValues.put(IMPORTANCE, success.importance)
        contentValues.put(DESCRIPTION, success.description)
        contentValues.put(DATE_STARTED, success.dateStarted)
        contentValues.put(DATE_ENDED, success.dateEnded)
        return contentValues
    }


    fun clear() {
        sqLiteDatabase?.delete(TB_NAME_SUCCESSES, null, null)
        sqLiteDatabase?.delete(TB_NAME_IMAGES, null, null)
    }
}
